src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -67,7 +72,12 @@ public EnumConverter(EnumConverterOptions converterOptions, JsonNamingPolicy? na
                     namingPolicy == null
                         ? JsonEncodedText.Encode(name, encoder)
                         : FormatEnumValue(name, encoder));
+
+                if (namingPolicy != null)
+                    _sourceNameCache.TryAdd(FormatEnumValue(name, encoder), name);
```

Comment:
This won't work for `Flags` enums and you also don't want to cache all the possible combinations since that would cause a security vulnerability; see https://github.com/dotnet/runtime/pull/36726#discussion_r432215725.